4 задание пункт 2

cryptojopa opened the pull request:
1) Создал DTO объекты(BankAccountDto, CategotyDto, TransactionDto, UserDto)
2) Реализовал валидацию для DTO объектов

cryptojopa committed:
4 задание
1) Создал DTO объекты(BankAccountDto, CategotyDto, TransactionDto, UserDto)
2) Реализовал валидацию для DTO объектов

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Objects;
+
+/**
+ * Объект банковского аккаунта
```
DTO != объект класса сущности (в javadoc'е лучше переименовать или в DTO, или что-то вроде объекта передачи данных)
Дополнительно почитать:
1) https://www.baeldung.com/java-dto-pattern
2) https://www.linkedin.com/pulse/difference-between-entity-dto-what-use-instead-babar-shahzad

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Объект банковского аккаунта
+ */
+@Getter
```
Cчитается хорошей практикой делать dto неизменяемыми (т.к. они по сути применяются только для передачи данных, и не несут в себе бизнес логики). С учетом того, что используем 17 версию java, можно просто использовать record'ы, либо изменить аннотации ломбока на соответствующие.
Дополнительно почитать: https://www.baeldung.com/java-entity-vs-dto#:~:text=Making%20DTOs%20immutable%20is%20a,final%20and%20not%20implementing%20setters.

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -0,0 +1,40 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class BankAccountDto {
+    @NotBlank(message = "Поле \"Название банка\" не может быть пустым")
```
Не название банка, а название банковского счета

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -0,0 +1,40 @@
+@AllArgsConstructor
+public class BankAccountDto {
+    @NotBlank(message = "Поле \"Название банка\" не может быть пустым")
+    @Size(min = 3, max = 15, message = "Длина названия банка должна содержать от 3 до 15 символов")
```
Аналогично

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -0,0 +1,40 @@
+    private String name;
+
+    @NotNull(message = "Поле \"Сумма счета\" не может быть пустым")
+    @DecimalMin(value = "0.0", message = "Сума счета не может быть отрицательной")
```
сума -> сумма

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -0,0 +1,40 @@
+    @DecimalMin(value = "0.0", message = "Сума счета не может быть отрицательной")
+    private BigDecimal amount;
+
+    @Override
```
Название банковского счета не является глобально уникальным идентификатором: у двух пользователей могут существовать счета с одинаковыми названиями, но в то же время, это не будут одни и те же счета.

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/CategoryDto.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.Objects;
+
+/**
+ * Объект категории
```
Аналогично замечанию про javadoc DTO

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/CategoryDto.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Объект категории
+ */
+@Getter
```
Аналогично замечанию про неизменяемость DTO

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/CategoryDto.java:
```diff
@@ -0,0 +1,36 @@
+    @Size(min = 3, max = 15, message = "Имя категории должно содержать от 3 до 15 символов")
+    private String name;
+
+    @Override
```
Аналогично комментарию с equals & hashcode банковского счета

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Objects;
+
+/**
+ * Объект транзакции
```
Аналогично замечанию про javadoc DTO

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -0,0 +1,43 @@
+ * Объект транзакции
+ */
+@Getter
+@Setter
```
Аналогично замечанию про неизменяемость DTO

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -0,0 +1,43 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class TransactionDto {
+    @NotNull(message = "Поле \"Имя категории\" не может быть пустым")
```
Не имя, а id

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -0,0 +1,43 @@
+    @NotNull(message = "Поле \"Имя категории\" не может быть пустым")
+    private Long categoryId;
+
+    @NotNull(message = "Поле \"Имя счета\" не может быть пустым")
```
Аналогично

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -0,0 +1,43 @@
+    private Long bankAccountId;
+
+    @NotNull(message = "Поле \"Сумма транзакции\" не может быть пустым")
+    @DecimalMin(value = "0.0", message = "Сума транзакции не может быть отрицательной")
```
сума -> сумма

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -0,0 +1,43 @@
+    @DecimalMin(value = "0.0", message = "Сума транзакции не может быть отрицательной")
+    private BigDecimal amount;
+
+    @Override
```
Аналогично, выбраны неверные поля для equals & hashcode

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/UserDto.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Objects;
+
+/**
+ * Объект пользователя
```
Аналогично замечанию про javadoc DTO

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/UserDto.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Объект пользователя
+ */
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
Аналогично замечанию про неизменяемость DTO

ivankrn reviewed (commented):
Есть замечания

cryptojopa committed:
4 задание(Правки)
1) Убрал Setter Для каждого DTO
2) Изменил javadoc для каждого DTO
3) Для BankAccountDto, TransactionDto, CategoryDto добавил поле id
4) Для BankAccountDto, TransactionDto, CategoryDto переопределил equals  и hashCode
5) Исправил грамматические ошибки в сообщениях
6) В BankAccount добавил поле name

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/database/model/BankAccount.java:
```diff
@@ -26,6 +26,8 @@ public class BankAccount {
     @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "bank_account_id_seq")
     @Setter(AccessLevel.NONE)
     private Long id;
+    @Column(nullable = false, unique = true)
```
Аналогично предыдущему замечанию про неуникальность имени
![image](https://github.com/ivankrn/finance-tracker-project/assets/102463202/66306f08-a2f3-48df-8bf7-c24138abfab6)

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -4,37 +4,38 @@
+    @NotNull(message = "Поле \"ID банковского аккаунта\" не может быть пустым")
+    private long id;
+
+    @NotBlank(message = "Поле \"Название банковского аккаунта\" не может быть пустым")
+    @Size(min = 3, max = 15, message = "Длина названия банковского счета должна содержать от 3 до 15 символов")
```
Так все же банковский аккаунт или счет?

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/BankAccountDto.java:
```diff
@@ -4,37 +4,38 @@
+ * DTO банковского аккаунта
  */
 @Getter
-@Setter
```
Если решил не использовать record'ы, то проще использовать аннотацию Lombok'а @Value
https://projectlombok.org/features/Value

ivankrn commented on src/main/java/ru/naumen/enterprisejavacourse/financetracker/dto/TransactionDto.java:
```diff
@@ -5,39 +5,40 @@
     private Long categoryId;
 
-    @NotNull(message = "Поле \"Имя счета\" не может быть пустым")
+    @NotNull(message = "Поле \"ID банковского аккаунта\" не может быть пустым")
```
У нас банковские счета, а не аккаунты

cryptojopa committed:
4 задание(Правки 2)
1) Поменял текст в сообщениях
2) Добавил аннотацию @Value ко всем DTO
3) Удалил методы equals и hashCode за ненадобностью
4) Пометил @EqualsAndHashCode.Include поля, которые будут использоваться при генерации equals и hashCode

cryptojopa committed:
4 задание(Правки 3)
1) Добавил аннотацию @EqualsAndHashCode со свойством onlyExplicitlyIncluded = true

ivankrn reviewed (approved):
Можно мерджить
